### DISCUSSION TITLE
Just want to run an x86 assembled program in 86box

### DISCUSSION THREAD

**Discussion by voidstar78** [86Box/86Box#1473] [Q&A] [answered]
I'm trying to get a bare-bones "self booting" program going on the IBM PC.

For reference: Using cc65 (6502 C compiler), I can target a Commodore PET.  The PET "boots" to BASIC, but has enough of a kernel to support saving and loading programs to disk (or tape).  So I can compile a 6502 target program (built for the PET system) and load it from tape, and have it executed on the PET.

Conceptually, shouldn't a similar sequence be possible on the IBM PC and 86box?   When I start 86box with IBM PC 1981 (8088) I get a BASIC prompt.   If I type save "test"  I get audio noise that I assume is modeling the binary data being audibly saved - although I'm not sure where it is saving to (should be a virtual cassette) and how to load it back (after restarting 86box).     After typing in a BASIC program, is there a command for saving that program to disk?   (for Commodore PET 4.0, they had a DSAVE command).

I have a simple program that doesn't require any disk access - it just polls the keyboard and POKEs things directly to the text-video.  I can use WATCOM C compiler to compile to the  DOS-16bit target, and to a .COM executable.    I have my .COM program, now how do I get 86Box to run it?  (or the original IBM PC to load and execute it?)

**Comment by bachimusprime** [comment, marked as answer]
You'll wanna use something like WinImage to create a floppy disk image formatted with a FAT file system. Then just inject your .COM executable into the floppy image, save, and load the image into 86Box after you've booted into DOS. Then run your program from the command line.

**Reply by voidstar78** [reply]
Thanks, WinImage is perfect - I created a 360K disk image,  "injected" my COM file (created by WATCOM C - 8K COM file just to print Hello!), and was able to run it like you said.    That works for like a late 1980's 386 Machine.

Then I adjusted 86Box to an 8088 IBM PC [1981] Machine, with no disk drives, so that it boots to ROM BASIC.   I should be able to load my .COM from a cassette recording.   I tried the LOAD command, and it makes a tone like it is emulating loading data from an audio source - but it's still not working (I still have to understand more how the emulator is modeling that LOAD command).    

In the Apple ][ world there was the CATALOG command and BRUN, i.e. it's ROM BASIC did have basic disk drive support (similar to the Commodore 4.0 BASIC, it added a DLOAD command) -- those were either '79 or '80 machines.  So I'm assuming the ROM BASIC in the IBM PC of '81 had some sort of disk support, but perhaps not.   But it does have a SAVE/LOAD command, I'm just not sure what it is saving to and loading from yet - I'll try to dig into some early IBM PC tech manual.   The 86BOX is claiming "Version C1.00" in the ROM that I have, I might need a C1.10 .

Think I'm close, but have to head out for a few hours.

**Reply by bachimusprime** [reply]
Oh, right, 86Box currently cannot load cassette images. You'll have to use floppies for now, unfortunately.

**Reply by voidstar78** [reply]
I see.  Ok the PCE-IBMPC emulator lets me do **LOAD"ldcass**  to load the IBM diagnostic tool from cassette (as described in the IBM user manual).  That IBM manual also described the C1.0 (cassette BASIC),  D1.0  (disk BASIC), and A1.0  (advanced BASIC) versions of BASIC.  At least for ROM BASIC C1.0, the manual calls out that it has no disk support.   I'm just imagining a situation of like: what if all the disk drive motors of the world were defunct, or there was no power to operate them -- how could I still make this machine useful.   You can almost crank a cassette by hand for as slow as they operate.   But I get your point: whether it is loaded by tape or disk, fundamentally the .COM program I construct will be the same  (and that's how it is on the PET and Apple ][, my same binary loads and run whether loaded by tape or from a disk -- on the PET I have an SD chip to emulate the whole disk drive).   Cool, and I have WATCOM C compiling my C code to make a .COM file - now I just need to adjust a few PEEK/POKE (to monitor keyboard and draw to the text screen), and figure out the MODE instructions that are placing the machine into 40x25 mode.

NOTE the PCE-IBMPC emulator has cassette BASIC C1.10.  The manual for that is here:  https://archive.org/details/IBMBASICAV1.10Manual/page/n1/mode/2up   
C1.20 was found on the IBM PCjr.  And apparently there also existed an IBM BASIC Compiler (that could compile BASIC but had some slight differences to its commands).  On Page 2-14, I learned something I didn't know: you can use ALT keys to quickly type BASIC commands, like ALT-P to enter "PRINT", etc.

I'm not so interested in BASIC - but I am interested in disassembly of BASIC and replicating the system calls it does to poll keyboard, print characters, change video modes, etc.

It's fascinating to me that they barely had computers in 1981, and yet they managed to create a 500+ page user manual with careful illustrations on how to use their products (all prepared BEFORE the launch of their products).  In a world before computers, they sure got a lot of stuff done.  Now that we have computers, it seems we're so multi-tasked and distracted to really focus on quality work.    For example, the PDF manual isn't yet searchable - but thankfully, the 500+ page manual has an Index, so to find info about PEEK, it tells me to go to page 4-205 and right there it is.

Anyhow, I'm on my way to what I wanted to get done with the 1981 IBM PC - and WinImage is key, on how to transfer my Windows 10 produced content over to a floppy disk image and try it out :)
